torchx/schedulers/kubernetes_mcad_scheduler.py

```diff
@@ -0,0 +1,1121 @@
+    """
+    Invokes ``lower`` on thes string and removes all
+    characters that do not satisfy ``[a-z0-9]`` pattern.
+    This method is mostly used to make sure kubernetes scheduler gets
+    the job name that does not violate its validation.
+    """
+    if data.startswith("-"):
+        data = data[1:]
+    pattern = r"[a-z0-9\-]"
+    return "".join(re.findall(pattern, data.lower()))
+
+
+def get_port_for_service(app: AppDef) -> str:
+    # Initialize port to default
+    port = "29500"
+
+    # look for rdzv_endpoint
+    rdzv = "--rdzv_endpoint"
+    separator = ":"
+    space = " "
+    for role_idx, role in enumerate(app.roles):
+        if rdzv in role.args[1]:
+            start_index = role.args[1].find(rdzv)
+
+            # find first instance of : after rdzv endpoint
+            port_index = role.args[1].find(separator, start_index) + 1
+            end_index = role.args[1].find(space, port_index)
+            port = role.args[1][port_index:end_index]
+
+    return port
+
+
+def app_to_resource(
+    app: AppDef,
+    namespace: str,
+    service_account: Optional[str],
+    image_secret: Optional[str],
+    priority: Optional[int] = None,
+) -> Dict[str, Any]:
+    """
+    app_to_resource creates a AppWrapper/MCAD Kubernetes resource definition from
+    the provided AppDef. The resource definition can be used to launch the
+    app on Kubernetes.
+
+    """
+
+    genericitems = []
+
+    unique_app_id = cleanup_str(make_unique(app.name))
+    for role_idx, role in enumerate(app.roles):
+        for replica_id in range(role.num_replicas):
+            values = macros.Values(
+                img_root="",
+                app_id=unique_app_id,
+                replica_id=str(replica_id),
+                rank0_env=f"MCAD_{cleanup_str(app.roles[0].name)}_0_HOSTS".upper().replace(
+                    "-", ""
+                ),
+            )
+
+            if role_idx == 0 and replica_id == 0:
+                values.rank0_env = "TORCHX_RANK0_HOST"
+            name = cleanup_str(f"{unique_app_id}-{replica_id}")
+            replica_role = values.apply(role)
+            if role_idx == 0 and replica_id == 0:
+                replica_role.env["TORCHX_RANK0_HOST"] = "localhost"
+
+            pod = role_to_pod(
+                name,
+                unique_app_id,
+                namespace,
+                replica_role,
+                service_account,
+                image_secret,
+            )
+            pod.metadata.labels.update(pod_labels(app, role_idx, role, replica_id))
+
+            genericitem: Dict[str, Any] = {
+                "replicas": 1,
+                "generictemplate": pod,
+            }
+            # TODO: retry support here
+            if role.max_retries > 0:
+                genericitem["maxRetry"] = role.max_retries
+                genericitem["policies"] = RETRY_POLICIES[role.retry_policy]
+                msg = f"""
+Role {role.name} configured with restarts: {role.max_retries}.
```

Comment:
nit: add comment about why we're warning about this "MCAD doesn't support retries. Role ... configured..."

Reply:
Updated message in commit: bcb0ebe432db59f8ecca3c0ac77bd8bdc2d5a83a